File: swift/CLDController.swift
public class CLDController {
    var callbacks = [(CLDContext) -> Void]()
    var context: CLDContext
    var functions = [(CLDContext) -> CLDContext]()
    var isProcessingQueue = false
    var queue = [CLDContext]()

    public init(_ context: CLDContext) {
        self.context = context
    }

    func executeFunctions() {
        /**/print("ИГР executeF-01 queue.count:", queue.count);
        let c = queue.removeFirst()
        context.recentField = c.recentField
        context.setField(c.recentField, c.fieldAny(c.recentField))
      
        for f in functions {
            let ctx = f(context)
            if ctx.recentField != "none" {
                queue.append(ctx)
            }
        }
      
        reportContext()
    }

    func processQueue() {
        // Prevent recursion.
        if isProcessingQueue {
            return
        }
      
        isProcessingQueue = true
      
        while (queue.count > 0) {
            executeFunctions()
        }
      
        isProcessingQueue = false
    }

    public func registerCallback(_ cb: @escaping (CLDContext) -> Void) {
        callbacks.append(cb)
    }

    public func registerFieldCallback(
        _ fieldName: String,
        _ cb: @escaping (CLDContext) -> Void
    ) {
        callbacks.append({ c in
            if c.recentField == fieldName {
                cb(c)
            }
        })
    }

    public func registerFunction(_ f: @escaping (CLDContext) -> CLDContext) {
        functions.append(f)
    }

    func reportContext() {
        for cb in callbacks {
            cb(context)
        }
    }

    public func set(
        _ fieldName: String,
        _ value: Any
    ) {
        var c = context
        c.setField(fieldName, value)
        c.recentField = fieldName
        queue.append(c)
        processQueue()
    }
}

File: javascript/CLDController.js
/**
 * This file is part of Cross-language dialect:
 *     https://github.com/OGStudio/cross-language-dialect
 * License: CC0
 * Version: 1.0.1
 */

function CLDController(context) {
    this._construct = function() {
        this.callbacks = [];
        this.context = context;
        this.functions = [];
        this.isProcessingQueue = false;
        this.queue = [];
    };
    this._construct();
 
    this.executeFunctions = function() {
        let c = this.queue.shift();
        this.context.recentField = c.recentField;
        this.context.setField(c.recentField, c.field(c.recentField));

        for (let i in this.functions) {
            let f = functions[i];
            let ctx = f(this.context.selfCopy());
            if (ctx.recentField != "none") {
                queue.push(ctx);
            }
        }

        reportContext();
    };

    this.processQueue = function() {
        // Prevent recursion.
        if (this.isProcessingQueue) {
            return;
        }

        this.isProcessingQueue = true;

        while (this.queue.length > 0) {
            this.executeFunctions();
        }

        this.isProcessingQueue = false;
    };

    this.registerCallback = function(cb) {
        this.callbacks.push(cb);
    };

    this.registerFieldCallback = function(fieldName, cb) {
        this.callbacks.push((c) => {
            if (c.recentField == fieldName) {
                cb(c);
            }
        });
    };

    this.registerFunction = function(f) {
        this.functions.push(f);
    };

    this.reportContext = function() {
        for (let i in this.callbacks) {
            let cb = this.callbacks[i];
            cb(this.context);
        }
    };

    this.set = function(fieldName, value) {
        let c - this.context.selfCopy();
        c.setField(fieldName, value);
        c.recentField = fieldName;
        this.queue.push(c);
        this.processQueue();
    };
}
